x/feeshare/post/post.go

```diff
@@ -0,0 +1,160 @@
+		return next(ctx, tx, simulate, success)
+	}
+
+	// Parse the transactions to FeeTx to get the total fees paid
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+	if feeTx.GetFee().Empty() || feeTx.GetFee().IsZero() {
+		return next(ctx, tx, simulate, success)
+	}
+
+	err = fsd.FeeSharePayout(ctx, feeTx.GetFee(), params.DeveloperShares, params.AllowedDenoms)
+	if err != nil {
+		return ctx, errorsmod.Wrapf(sdkerrors.ErrLogic, err.Error())
+	}
+
+	return next(ctx, tx, simulate, success)
+}
+
+// FeeSharePayout takes the total fees paid for a transaction and
+// split these fees equaly between all the contacts involved in the
+// transaction based on the module params.
+func (fsd FeeSharePayoutDecorator) FeeSharePayout(ctx sdk.Context, txFees sdk.Coins, devShares sdk.Dec, allowedDenoms []string) (err error) {
+	executedContracts, found := fsd.wasmKeeper.GetExecutedContractAddresses(ctx)
+	if !found {
+		return err
+	}
+	contractAddresses := executedContracts.ContractAddresses
+	if len(contractAddresses) == 0 {
+		return err
+	}
+
+	withdrawerAddrs, err := GetWithdrawalAddressFromContract(ctx, contractAddresses, fsd.feesharekeeper)
+	if err != nil {
+		return err
+	}
+	if len(withdrawerAddrs) == 0 {
+		return err
+	}
+
+	feeToBePaid := CalculateFee(txFees, devShares, len(withdrawerAddrs), allowedDenoms)
+	if feeToBePaid.IsZero() {
+		return err
+	}
+
+	// pay the fees to the withdrawer addresses
+	for _, withdrawerAddrs := range withdrawerAddrs {
+		err = fsd.bankKeeper.SendCoinsFromModuleToAccount(ctx, authtypes.FeeCollectorName, withdrawerAddrs, feeToBePaid)
+		if err != nil {
+			return err
+		}
+		err := ctx.EventManager().EmitTypedEvent(
+			&feeshare.FeePayoutEvent{
+				WithdrawAddress: withdrawerAddrs.String(),
+				FeesPaid:        feeToBePaid,
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// Iterate the contract addresses and get the
+// withdrawer address from the module store
+func GetWithdrawalAddressFromContract(ctx sdk.Context, contractAddresses []string, fsk FeeShareKeeper) ([]sdk.AccAddress, error) {
+	var withdrawerAddrs []sdk.AccAddress
+
+	for _, contractAddr := range contractAddresses {
+		parsedContractAddr, err := sdk.AccAddressFromBech32(contractAddr)
+		if err != nil {
+			return nil, err
+		}
+
+		shareData, hasfeeshare := fsk.GetFeeShare(ctx, parsedContractAddr)
+
+		if !hasfeeshare {
+			continue
+		}
+
+		withdrawerAddr := shareData.GetWithdrawerAddr()
+		if withdrawerAddr != nil && !withdrawerAddr.Empty() {
+			withdrawerAddrs = append(withdrawerAddrs, withdrawerAddr)
+		}
+	}
+
+	return withdrawerAddrs, nil
+}
+
+// CalculateFee takes the total fees paid for a transaction and split
+// these fees equaly between all number of pairs considering allwoedDenoms
+func CalculateFee(fees sdk.Coins, govPercent sdk.Dec, pairs int, allowedDenoms []string) sdk.Coins {
```

Comment:
govPercent should instead use devShares to be consistent.
pairs could be renamed to `numOfdevs` or `devCount`

Reply:
calculate fees should use `number of contracts` interacted for `pairs` instead of pair counts.
If a transaction interacted with 5 contracts and only 1 of them has fee share withdrawal address registered, the dev should only get 10% of the fees instead of 50%. (assuming dev shares is 0.5)